### ISSUE TITLE
The implementation of DCI without using macro-programming

### ISSUE THREAD

**Issue by ajukraine** [ciscoheat/haxedci#2]
The way, you provide an environment for DCI, is nice and self-explanatory. The only tough thing is IDE support -- it's often the problem, when using Haxe macros.

Here is the implementation of DCI using standard Haxe syntax (without macros):

```haxe
package;

class Main {
	static public function main() {
		var acc1 = new AccountContext([]);

		acc1.deposit(10);
		acc1.deposit(15.5);
		acc1.withdraw(7);

		var acc2 = new AccountContext([]);

		var moneyTransfer = new MoneyTransferContext(acc1, acc2);
		moneyTransfer.transfer(8.5);

		trace(acc1.getBalance());
		trace(acc2.getBalance());

		var accountMock = {
			deposit: function(amount: Float) {
				trace("Mocking");
			}
		};

		var moneyTransfer2 = new MoneyTransferContext(acc1, accountMock);
		moneyTransfer2.transfer(5);
	}
}

//
// AccountingEntry.hx

class AccountingEntry {
	public function new() {}

	public var amount(default, default) : Float;
	public var message(default, default) : String;
}

//
// AccountContext.hx

class AccountContext {
	var bookKeeper: BookKeeper;

	public function new(entries) {
		this.bookKeeper = entries;
	}

	public function getBalance() : Float {
		return bookKeeper.getBalance();
	}

	public function deposit(amount : Float) {
		trace("Depositing");
		bookKeeper.addEntry("Depositing", amount);
	}

	public function withdraw(amount : Float) {
		bookKeeper.addEntry("Withdrawing", -amount);
	}
}

abstract BookKeeper(Array<AccountingEntry>) from Array<AccountingEntry> {

	public function getBalance() : Float {
		return Lambda.fold(this, function(a, b) { return a.amount + b; }, 0.0);
	}

	public function addEntry(message : String, amount : Float) {
		var entry = new AccountingEntry();
		entry.amount = amount;
		entry.message = message;

		this.push(entry);
	}
}

//
// MoneyTransferContext.hx

class MoneyTransferContext {
	var sourceAccount : SourceAccount;
	var destinationAccount : DestinationAccount;

	public function new(source : SourceAccountActor, destination : DestinationAccountActor) {
		this.sourceAccount = source;
		this.destinationAccount = destination;
	}

	public function transfer(amount : Float) {
		sourceAccount.transferTo(destinationAccount, amount);
	}
}

typedef SourceAccountActor = {
	function getBalance() : Float;
	function withdraw(amount: Float) : Void;
}

typedef DestinationAccountActor = {
	function deposit(amount: Float) : Void;
}

abstract SourceAccount(SourceAccountActor) from SourceAccountActor {
	public function transferTo(destinationAccount : DestinationAccount, amount : Float) {
		if (this.getBalance() < amount) {
			throw "Insufficient funds";
		}

		this.withdraw(amount);
		destinationAccount.deposit(amount);
	}
}

@:forward(deposit)
abstract DestinationAccount(DestinationAccountActor) from DestinationAccountActor
{ }
```

Main advantage: built-in Haxe support
Main disadvantages: lacks DCI-specific language (which is provided by your macros library)

Do you agree? How you compare these two approaches? I'm very eager to get your answers and thoughts.

**Comment by ciscoheat** [comment]
Hi, glad to see your interest in DCI! I was using Abstract types first, but decided to change to the macro approach because of a few reasons. For example, using Abstracts you need to practically pass the Context as parameters to the RoleMethod, as you see in `transferTo`. With the macro approach you have access to other Roles directly. Having to pass Roles around in methods kind of defeats the purpose with RoleMethods, they should be used to transport information relevant to the messages passed in the Context. As in a simplified Restaurant example: Food is created by the Chef and is passed to the Waiter and finally to the Customer. That's a more valid usage of parameters in RoleMethods.

Another reason is encapsulation. The Roles should be private to their Context because they don't make sense anywhere else. They can be made private for the module of course, but I think the syntax for the macro version is more readable, and that's one of the foremost goals with DCI - readable code with the least amount of surprises. Then there are some other small things with Abstracts like having to use `@:forward`. Small things that need to be kept in memory which will hinder usability.

Finally with macros you can enforce DCI rules in a way that will help the programmer. You can prevent duplicate names, usage of `this` in RoleMethods, Role aliasing and other things that can cause surprises for a programmer reading the code.

So to summarize, I see too made advantages with the macro approach not to use it. :) IDE support is pretty good with the latest version, it works most of the time in both Sublime and Flashdevelop, and the 3.2 version seems to have made it even better (it had some problems in 3.1).

**Comment by ajukraine** [comment]
@ciscoheat Actually you're right. I've just tried `haxedci` in my sandbox -- and it's nicely supported by Sublime. And in general, for complex use cases it's better for role object to have access to other objects through implicit context (kind of closure for roles). Also it makes perfect sense for `role` to be 1-st class citizen of the application.

I appreciate your efforts in order to create Haxe environment for DCI paradigm. At the moment I don't better implementations of DCI (event Ruby/Smalltalk examples are cluttered with syntax overhead). Great job.

**Comment by ciscoheat** [comment]
Thank you for your kind words. :) I'm going to use this library for a large project this year, so I will keep it up-to-date and in sync with the discussions on the DCI list. If you have some DCI code for show, please drop me a message! I'm very interested in others creations.
